File: my-app/src/types/sharing.ts
export type TopicStatus = "voting" | "confirmed" | "completed";

export interface TopicListItem {
  id: number;
  title: string;
  status: TopicStatus;
  presenters: string | null;
  session_number: number | null;
  shared_at: string | null;
  has_materials: boolean;
  vote_count: number;
  current_user_voted: boolean;
  is_student_submitted: boolean;
  submitted_by_name: string | null;
}

export interface TopicListResponse {
  items: TopicListItem[];
  total_votes: number;
}

export interface TopicMaterialsResponse {
  topic_id: number;
  title: string;
  materials_content: string;
}

export interface VoteResponse {
  vote_count: number;
}

export interface AdminTopicListItem extends TopicListItem {
  materials_content: string | null;
}

export interface AdminTopicListResponse {
  items: AdminTopicListItem[];
}

export interface TopicCreateRequest {
  title: string;
  class_id: number;
  status?: TopicStatus;
  presenters?: string | null;
  session_number?: number | null;
  shared_at?: string | null;
  materials_content?: string | null;
}

export interface TopicUpdateRequest {
  title?: string;
  status?: TopicStatus;
  presenters?: string | null;
  session_number?: number | null;
  shared_at?: string | null;
  materials_content?: string | null;
}

export interface TopicSuggestRequest {
  title: string;
}

File: backend/schemas/sharing.py
import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel

TopicStatusType = Literal["voting", "confirmed", "completed"]

MAX_VOTES_PER_STUDENT = 3


# --- Requests ---

class TopicCreateRequest(BaseModel):
    title: str
    class_id: uuid.UUID
    status: TopicStatusType = "voting"
    presenters: str | None = None
    session_number: int | None = None
    shared_at: datetime | None = None
    materials_content: str | None = None


class TopicUpdateRequest(BaseModel):
    title: str | None = None
    status: TopicStatusType | None = None
    presenters: str | None = None
    session_number: int | None = None
    shared_at: datetime | None = None
    materials_content: str | None = None


class TopicSuggestRequest(BaseModel):
    title: str


# --- Responses ---

class TopicListItem(BaseModel):
    id: uuid.UUID
    title: str
    status: TopicStatusType
    presenters: str | None
    session_number: int | None
    shared_at: datetime | None
    has_materials: bool
    vote_count: int
    current_user_voted: bool
    is_student_submitted: bool
    submitted_by_name: str | None


class TopicListResponse(BaseModel):
    items: list[TopicListItem]
    total_votes: int


class TopicMaterialsResponse(BaseModel):
    topic_id: uuid.UUID
    title: str
    materials_content: str


class VoteResponse(BaseModel):
    vote_count: int


# --- Admin responses ---

class AdminTopicListItem(TopicListItem):
    materials_content: str | None


class AdminTopicListResponse(BaseModel):
    items: list[AdminTopicListItem]
